### DISCUSSION TITLE
Network templates

### DISCUSSION THREAD

**Discussion by WilliamDEdwards** [netbox-community/netbox#6242] [Ideas]
When creating a new network, the deployment steps are usually the same:

* Get free vLAN ID
* Add /128 and /32 to router 1 device (with DNS name like: `rtr1-{vlan_id}.ourdomain.net`)
* Add /128 and /32 IP to router 2 device (with DNS name like: `rtr2-{vlan_id}.ourdomain.net`)
* Get free VID (see: https://github.com/netbox-community/netbox/issues/6235)
* Add VRRP6 and VRRP IP to router 1 device and router 2 device (with DNS name like: `gw-{vlan_id}.ourdomain.net`)

It would be useful to create all objects needed to deploy a new network at the click of a button. Maybe based on templates, because not every network is exactly the same.

I haven't thought about this enough to propose a structure, but I guess it'd come down to:

* Create a new template model.
* The user should create a template. The user may add vLAN, prefix and IP address objects to the template. These objects will be created when a network is created from the template. In some object fields, like the IP address's DNS name, variable substitution (like `vlan_id`) should be possible.
* The user can then create the network, which is not an object itself, but the result of a collection of objects, based on one of their templates.

Although the templates concept would allow for some flexibility, I feel like this may be too dependent on specific infrastructure setups to be useful for all NetBox users.

What are other's approaches to this?

**Comment by candlerb** [comment]
The approach I suggest is to write a [custom script](https://netbox.readthedocs.io/en/stable/additional-features/custom-scripts/).  It can perform all the steps you listed, exactly to your requirements, and do them safely within a single database transaction.

You can see some examples [here](https://github.com/netbox-community/reports/tree/master/scripts).

**Reply by WilliamDEdwards** [reply]
Thanks for your reply. Yep, that's the obvious solution. I created this issue to see if maybe it would be useful to natively implement this in NetBox (but as I said, I feel like this may be too dependent on specific infrastructure setups to be useful for all NetBox users).
